Initial version of HTTP sink exporter

owais opened the pull request:
This is a test exporter that allows clients to "watch" for spans they might be interested in. This is very useful for end-to-end tests where we want to test real world applications and ensure data reaches the collector as expected.

The exporter exposes a HTTP endpoint. Clients can make GET requests to this endpoint and specify the characteristics of spans they are interested in using URL query params. Collector will then block the request until it detects the specified spans. Once the spans are detected, they are serialized to JSON (using jaeger proto) and returned back to the client. Multiple clients can watch for the same spans at the same time. The export does not hold anything in memory before a watch is started or after a watch ends. Clients must start the watch first, generate spans and assert the watch result.

All this can be done with the file or logging exporter as well but this dramatically simplifies writing such tests especially when this needs to be done for multiple libraries across different languages.

Example test that leverages this to test a simple but real app instrumented with otel: https://github.com/signalfx/splunk-otel-python/blob/main/tests/integration/test_simple.py#L24-L47 

The implementation is very basic, code is rough and a lot of tests are lacking. I'll improve this over time and if it makes sense, try to migrate it to Otel Contrib but for now this would make life a lot easier for at least a couple of our (instrumentals) projects.

Perhaps down the line, we can permanently host a collector in Splunk lab with this exporter enabled to allow very quick and easy manual and automated testing (even from CI). May be I'll try this for the upcoming hackathon. 

### Example and explanation 
Recorded an example here: https://www.loom.com/share/16256b778d9a40a297a2212692422d0b

rmfitzpatrick commented on internal/exporter/pollingexporter/README.md:
```diff
@@ -0,0 +1,28 @@
+receives such spans or the request times out. Once the requested spans are detected, they're
+returned back to the client as JSON. 
+
+This exporter return data as [JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json)
```
```suggestion
This exporter returns data as [JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json)
```

It would be helpful to show an example request w/ params and the response imo.

owais replied:
Updated description with link: https://www.loom.com/share/16256b778d9a40a297a2212692422d0b

jrcamp commented:
Why jaeger instead of OTLP JSON? Presumably this would be useful for logs and metrics as well. Once https://github.com/open-telemetry/opentelemetry-collector/tree/main/internal/model is more mature can use it to serialize to any protocol.

owais commented:
It was a lot more work to use OTLP as collector doesn't expose the models. I could re-generate the proto definitions for OTLP have use them but even that was a lot of work at the time. The goal was to allow splunk-otel-python to use this for integration tests and Jaeger allowed me to get there fastest. 

We should definitely support OTLP when we can though. It might even be worth adding `?format=otlp|jaeger|zipkin|...` query param and let the client chose the format it wants.

owais commented:
@jrcamp @rmfitzpatrick @dmitryax PTAL when you can. Thanks

rmfitzpatrick commented on internal/exporter/pollingexporter/README.md:
```diff
@@ -0,0 +1,28 @@
+# Polling Exporter
```
Would something like "http testing" or "sanity" exporter be more accurate and denote this is for testing purposes only?

owais replied:
Yes polling isn't really a good name. I don't want to call it testing as that implies automated testing and this can be useful for manual tests as well but for sure that'd be better than polling )

owais replied:
Renamed to `httpsink`.

rmfitzpatrick commented on internal/exporter/pollingexporter/client.go:
```diff
@@ -0,0 +1,119 @@
+}
+
+func (c *client) response() ([]byte, error) {
+	// add support to filter by attributes and names
```
```suggestion
	// TODO: add support to filter by attributes and names
```

rmfitzpatrick commented on internal/exporter/pollingexporter/client.go:
```diff
@@ -0,0 +1,119 @@
+				}
+			}
+		case <-time.After(c.opts.timeout):
+			return nil, fmt.Errorf("timedout while waiting for spans")
```
```suggestion
			return nil, fmt.Errorf("timed out while waiting for spans")
```

rmfitzpatrick replied:
Could be helpful to report how many were received so this isn't all or nothing.

owais replied:
This could be useful but none of the clients (Splunk Otel Python and Splunk Otel JS) need it right now. We can definitely add it if a real user requests it. May be we can even make it a streaming response and returns spans as they arrive without waiting for all.

rmfitzpatrick commented on internal/exporter/pollingexporter/polling_exporter.go:
```diff
@@ -0,0 +1,110 @@
+
+	result, err := c.response()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
404 makes more sense than 500 to me, so maybe response should also include a status code?

owais replied:
done

rmfitzpatrick commented on internal/exporter/pollingexporter/polling_exporter.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(result)
```
Should a client be unregistered after its response is written?

owais replied:
Yes, I was holding that off for future enhancement but can do that now if you'd like.

rmfitzpatrick replied:
I'd think it'd be preferable to deferring (a loose end to manage)

owais replied:
done

emaderer commented:
@owais  can we close this PR for now?

Rauno56 commented:
Can we work towards merging it?
We needed to e2e test JS distro and are using polling exporter for it.

owais commented:
I have updated the PR and addressed @rmfitzpatrick's comments. It would be nice to get this in as two projects already depend on this for integration tests (splunk-otel-python and splunk-otel-js).
